File: packages/sdkwork-utils-go/result.go
package utils

type ResultValue[T any] struct {
	Value T
	Error string
	Ok    bool
}

func Ok[T any](value T) ResultValue[T] {
	return ResultValue[T]{Value: value, Ok: true}
}

func Err[T any](message string) ResultValue[T] {
	return ResultValue[T]{Error: message, Ok: false}
}

func IsOk[T any](result ResultValue[T]) bool {
	return result.Ok
}

func IsErr[T any](result ResultValue[T]) bool {
	return !result.Ok
}

func UnwrapOr[T any](result ResultValue[T], defaultValue T) T {
	if result.Ok {
		return result.Value
	}
	return defaultValue
}

func Map[T any, U any](result ResultValue[T], mapper func(T) U) ResultValue[U] {
	if !result.Ok {
		return Err[U](result.Error)
	}
	return Ok(mapper(result.Value))
}

File: packages/sdkwork-utils-typescript/src/result.ts
export type ResultValue<T> =
  | { ok: true; value: T }
  | { ok: false; error: string };

export function ok<T>(value: T): ResultValue<T> {
  return { ok: true, value };
}

export function err<T = never>(message: string): ResultValue<T> {
  return { ok: false, error: message };
}

export function isOk<T>(result: ResultValue<T>): result is { ok: true; value: T } {
  return result.ok;
}

export function isErr<T>(result: ResultValue<T>): result is { ok: false; error: string } {
  return !result.ok;
}

export function unwrapOr<T>(result: ResultValue<T>, defaultValue: T): T {
  return result.ok ? result.value : defaultValue;
}

export function map<T, U>(result: ResultValue<T>, mapper: (value: T) => U): ResultValue<U> {
  return result.ok ? ok(mapper(result.value)) : err(result.error);
}
